KVM HA Enhancements for Ceph RBD Primary Storage

dhslove opened the pull request:
### Description

This PR includes the features to provide high availability using Ceph RBD-based Primary Storage.

The KVM Host HA feature uses NFS Primary Storage. However, an additional storage consideration in KVM is Ceph RBD. This PR considers the following two to provide HA using RBD.

1. Must use the multi-monitor server feature of Ceph RBD.
2. The RBD-based HA function should be provided in the same way as the NFS-based HA.

This PR includes the feature to enter multiple monitor addresses separated by commas in the RADOS Monitor item when registering RBD Primary Storage.

And it includes adding code to provide HA based on RBD Primary Storage.

To support RBD-based HA, add RBDStoragePool class to KVMHABase and use it to perform heartbeat check and activity check in the same way as NFS. Similar to how NFS writes hb and ac files to a KVMHA directory, RADOS commands are used to put hb and ac files into the RBD pool registered as primary storage.

To perform this heartbeat and activity check, kvmheartbeat_rbd.sh and kvmvmactivity_rbd.sh files were added. This is to clearly distinguish roles from the same shell file for nfs.

This PR is necessary to support agent-free HA based on Ceph RBD like NFS, even if the related PR #4978 is merged.

The execution result of the function can be checked in the attached image.


<!--- Describe your changes in DETAIL - And how has behaviour functionally changed. -->

<!-- For new features, provide link to FS, dev ML discussion etc. -->
<!-- In case of bug fix, the expected and actual behaviours, steps to reproduce. -->

<!-- When "Fixes: #<id>" is specified, the issue/PR will automatically be closed when this PR gets merged -->
<!-- For addressing multiple issues/PRs, use multiple "Fixes: #<id>" -->
<!-- Fixes: # -->

<!--- ********************************************************************************* -->
<!--- NOTE: AUTOMATATION USES THE DESCRIPTIONS TO SET LABELS AND PRODUCE DOCUMENTATION. -->
<!--- PLEASE PUT AN 'X' in only **ONE** box -->
<!--- ********************************************************************************* -->

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial


### Screenshots (if appropriate):


### How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->



<!-- Please read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/main/CONTRIBUTING.md) document -->
RBD Multiple Monitor Support - Add RBD Primary Storage
![rbd-primary-add](https://user-images.githubusercontent.com/44049806/149456525-d500be8b-5f37-4f3a-94e8-de669330f3cf.png)

RBD Multiple Monitor Support - RBD Primary Storage List
![primarystorage-list](https://user-images.githubusercontent.com/44049806/149456606-c13979c3-93f5-4306-9f93-c9f1953ffbd2.png)

RBD Multiple Monitor Support - KVM Host Secret List
![secret-list](https://user-images.githubusercontent.com/44049806/149456815-48f6b43e-cee4-4d9c-af3d-b1b4e34a23ef.png)

RBD Multiple Monitor Support - VM Disk XML(Multi RBD Monitor)
![disk-xml](https://user-images.githubusercontent.com/44049806/149456927-bdd0072e-042c-4232-a4ac-05b33cf8ef74.png)

RBD Based KVM Host HA - One of Hosts Off
![host-list](https://user-images.githubusercontent.com/44049806/149457092-ab981e76-f93b-4e21-af73-c99b78332c18.png)

RBD Based KVM Host HA - Host Suspect State
![host-status](https://user-images.githubusercontent.com/44049806/149457147-a4750057-c439-40b0-a20f-9d744189f3e7.png)

RBD Based KVM Host HA - Host Down State
![host-down1](https://user-images.githubusercontent.com/44049806/149457209-4c0c3393-7b8c-4b5c-85c2-8b6c7f8e0b7c.png)

Dajeong-Park committed:
rbd multi monitor 지원

rbd multi monitor 지원

Dajeong-Park committed:
comma로 넣을 경우 userInfo 도 null 로 출력되는 현상 수정

comma로 넣을 경우 userInfo 도 null 로 출력되는 현상 수정

Dajeong-Park committed:
rbd일 때 uuid 에러로 생성로직 변경

rbd일 때 uuid 에러로 생성로직 변경

Dajeong-Park committed:
multi monitor인 경우 로직 변경

multi monitor인 경우 로직 변경

Dajeong-Park committed:
코드 정리

코드 정리

ycyun committed:
Replace multiple + and / symbols by - and _ to have URL-safe base64 going to the API

dhslove committed:
Merge pull request #91 from ycyun/main

RBD primary storage 연결시 + 및 / 기호를 -와 _ 로 변경하는 로직 개선

Commit:
Host HA : RBD Primary Storage에 대한 Host HA 기능 제공

Primary Storage가 RBD인 경우에  Host HA가 동작하도록 기능개선

Commit:
Host HA : RBD Primary Storage에 대한 Host HA 기능 제공

nfsstoragepool, rbdstoragepool 분리

Commit:
Host HA : RBD Primary Storage에 대한 Host HA 기능 제공

storagepool > pooltype 제거 및 메서드명변경(getRbdMonIpAddress)

Commit:
Host HA : RBD Primary Storage에 대한 Host HA 기능 제공

rbd pool auth 변경가능하도록 수정

Commit:
Host HA : RBD Primary Storage에 대한 Host HA 기능 제공

nfs, rbd Storage hb모니터링 함께 되도록 수정

dhslove committed:
Merge branch 'main' into rbd-primary-ha

sureshanaparti commented:
@blueorangutan package

blueorangutan commented:
@sureshanaparti a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 2208

sureshanaparti commented:
@blueorangutan test

wido commented:
This looks very interesting. I'll review this and get back. Give me a few days

blueorangutan commented:
<b>Trillian test result (tid-2879)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 33668 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr5862-t2879-kvm-centos7.zip
Smoke tests completed. 92 look OK, 0 have errors
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

wido commented:
May I ask why you are not just looking up if there is a Watcher on RBD images?

dhslove commented:
> 

Thanks your review. Wido.

The reason is 
1. To use the same method as NFS (modified time check), 
2. RBD Watcher continues to exist when the KVM host hangs (Watcher's inaccuracy in some cases).

GutoVeronezi commented on plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java:
```diff
@@ -243,6 +257,10 @@ public DataStore initialize(Map<String, Object> dsInfos) {
             }
+            if (multi) {
+                storageHost = storageHost + (hostPath.substring(0, hostPath.lastIndexOf("/")).replaceAll("/", ","));
+                hostPath = hostPath.substring(hostPath.lastIndexOf("/")+1);
```
```suggestion
                hostPath = hostPath.substring(hostPath.lastIndexOf("/") + 1);
```

dhslove replied:
@GutoVeronezi   I modified the code as you requested.

GutoVeronezi commented on plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java:
```diff
@@ -162,10 +168,18 @@ public DataStore initialize(Map<String, Object> dsInfos) {
+                    String multiHost = uriHost + (uriPath.substring(0, uriPath.lastIndexOf("/")).replaceAll("/", ","));
+                    String[] hostArr = multiHost.split(",");
+                    if (hostArr.length > 5) {
+                        throw new InvalidParameterValueException("RADOS monitor can support up to 5.");
```
```suggestion
                        throw new InvalidParameterValueException("RADOS monitor can support up to 5 hosts.");
```

dhslove replied:
@GutoVeronezi   I modified the code as you requested.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtFenceCommandWrapper.java:
```diff
@@ -47,20 +48,25 @@ public Answer execute(final FenceCommand command, final LibvirtComputingResource
             String logline = "No NFS storage pools found. No way to safely fence " + command.getVmName() + " on host " + command.getHostGuid();
             s_logger.warn(logline);
             return new FenceAnswer(command, false, logline);
+        }else if (rbdpools.size() == 0) {
```
```suggestion
        } else if (rbdpools.size() == 0) {
```

dhslove replied:
@GutoVeronezi   I modified the code as you requested.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtFenceCommandWrapper.java:
```diff
@@ -47,20 +48,25 @@ public Answer execute(final FenceCommand command, final LibvirtComputingResource
             s_logger.warn(logline);
             return new FenceAnswer(command, false, logline);
+        }else if (rbdpools.size() == 0) {
+            String logline = "No RBD storage pools found. No way to safely fence " + command.getVmName() + " on host " + command.getHostGuid();
```
We could use `String.format` here.

dhslove replied:
@GutoVeronezi    I modified the code as you requested.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/wrapper/LibvirtCheckVMActivityOnStoragePoolCommandWrapper.java:
```diff
@@ -43,9 +44,13 @@ public Answer execute(final CheckVMActivityOnStoragePoolCommand command, final L
-            final KVMHAVMActivityChecker ha = new KVMHAVMActivityChecker(nfspool, command.getHost().getPrivateNetwork().getIp(), command.getVolumeList(), libvirtComputingResource.getVmActivityCheckPath(), command.getSuspectTimeInSeconds());
+            final RbdStoragePool rbdpool = monitor.getRbdStoragePool(pool.getUuid());
+            String vmActivityCheckPath = "";
+            if (Storage.StoragePoolType.NetworkFilesystem == pool.getType())    vmActivityCheckPath = libvirtComputingResource.getVmActivityCheckPath();
+            else if (Storage.StoragePoolType.RBD == pool.getType())    vmActivityCheckPath = libvirtComputingResource.getVmActivityCheckPathRbd();
```
According to our [coding conventions](https://cwiki.apache.org/confluence/display/CLOUDSTACK/Coding+conventions),  blocks must follow the following layout:

```java
if (condition) {
    statements;
} else {
    statements;
}
```

dhslove replied:
@GutoVeronezi   I modified the code as you requested.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHABase.java:
```diff
@@ -181,6 +206,26 @@ protected String runScriptRetry(String cmdString, OutputInterpreter interpreter)
+            }
+            return sourceHostIP;
+        } catch (UnknownHostException e) {
+            s_logger.debug("Failed to get connection: " + e.getMessage());
```
We could pass the exception as parameter.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:
```diff
@@ -35,121 +35,221 @@
+                        storage = conn.storagePoolLookupByUUIDString(uuid);
+                        if (storage == null || storage.getInfo().state != StoragePoolState.VIR_STORAGE_POOL_RUNNING) {
+                            if (storage == null) {
+                                s_logger.debug(String.format("Libvirt storage pool [%s] not found, removing from HA list.", uuid));
```
```suggestion
                                s_logger.debug(String.format("Libvirt storage pool [%s] was not found, removing it from HA list.", uuid));
```

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:
```diff
@@ -35,121 +35,221 @@
+                            if (storage == null) {
+                                s_logger.debug(String.format("Libvirt storage pool [%s] not found, removing from HA list.", uuid));
+                            } else {
+                                s_logger.debug(String.format("Libvirt storage pool [%s] found, but not running, removing from HA list.", uuid));
```
```suggestion
                                s_logger.debug(String.format("Libvirt storage pool [%s] was found, but it is not running, removing it from HA list.", uuid));
```

dhslove replied:
@GutoVeronezi   I modified the code as you requested.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:
```diff
@@ -35,121 +35,221 @@
+                        Script cmd = createHeartBeatCommand(primaryStoragePool, hostPrivateIp, true);
+                        result = cmd.execute();
+
+                        s_logger.debug(String.format("The command (%s), to the pool [%s], has the result [%s].", cmd.toString(), uuid, result));
```
```suggestion
                        s_logger.debug(String.format("The command [%s], to the pool [%s], had the result [%s].", cmd.toString(), uuid, result));
```

dhslove replied:
@GutoVeronezi     I modified the code as you requested.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:
```diff
@@ -35,121 +35,221 @@
-                            s_logger.debug(String.format("Libvirt storage pool [%s] not found, removing from HA list.", uuid));
-                        } else {
-                            s_logger.debug(String.format("Libvirt storage pool [%s] found, but not running, removing from HA list.", uuid));
+        if(nfsstoragePool != null && !nfsstoragePool.isEmpty()) {
```
```suggestion
        if (nfsstoragePool != null && !nfsstoragePool.isEmpty()) {
```

dhslove replied:
I modified the code as you requested.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:
```diff
@@ -35,121 +35,221 @@
+        }
 
-                    s_logger.debug(String.format("The command (%s), to the pool [%s], has the result [%s].", cmd.toString(), uuid, result));
+        if (rbdstoragePool != null && !rbdstoragePool.isEmpty()) {
```
This code and the one above look quite similar, we could extract it to a method

dhslove replied:
@GutoVeronezi  OK, I made method 'run

DaanHoogland replied:
I agree with @GutoVeronezi and you could do more than just the `runHeartbeatToPool()`. The rest of the block can be parameterised with the synchronisation object, `nfsstoragePool` and `rbdstoragePool` respectively. `obj.getClass()` to get the type and instantiate the and get the object in the loop, or add the type ars parameter as well.

no blocker, just a suggestion!

sjyu1 committed:
Merge pull request #5 from ablecloud-team/rbd-primary-ha

Rbd primary ha

Commit:
HA Cloudstack 요청사항 수정

코딩규칙 및 요청사항 수정

sjyu1 committed:
HA Cloudstack 요청사항 수정

runHeartbeatToPool 메서드명 수정

dhslove committed:
Merge pull request #147 from sjyu1/rbd-primary-ha

HA Cloudstack 요청사항 수정

DaanHoogland commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:
```diff
@@ -35,121 +35,189 @@
+                if (storage == null) {
+                    s_logger.debug(String.format("Libvirt storage pool [%s] not found, removing from HA list.", uuid));
+                } else {
+                    s_logger.debug(String.format("Libvirt storage pool [%s] was found, but it is not running, removing it from HA list.", uuid));
```
not as strict on logging as @GutoVeronezi but one suggestion (making it only once)
```suggestion
                    s_logger.debug(String.format("Libvirt storage pool [%s] was found, but it is not running, removing it from the HA list.", uuid));
```

wido commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHABase.java:
```diff
@@ -181,6 +206,26 @@ protected String runScriptRetry(String cmdString, OutputInterpreter interpreter)
+            String[] hostArr = sourceHost.split(",");
+            String sourceHostIP = "";
+            for (String host : hostArr) {
+                String ipRegex = "(\\d{1,2}|1\\d\\d|2[0-4]\\d|25[0-5])\\.(\\d{1,2}|1\\d\\d|2[0-4]\\d|25[0-5])\\.(\\d{1,2}|1\\d\\d|2[0-4]\\d|25[0-5])\\.(\\d{1,2}|1\\d\\d|2[0-4]\\d|25[0-5])";
```
Does this support IPv6?

wido commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHABase.java:
```diff
@@ -181,6 +206,26 @@ protected String runScriptRetry(String cmdString, OutputInterpreter interpreter)
+                if (host.matches(ipRegex)) {
+                    sourceHostIP += host + ",";
+                } else {
+                    InetAddress addr = InetAddress.getByName(host);
```
Does this support IPv6?

wido commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAChecker.java:
```diff
@@ -71,6 +73,30 @@ public Boolean checkingHeartBeat() {
+            cmd.add("-i", getRbdMonIpAddress(pool._poolSourceHost));
+            cmd.add("-p", pool._poolMountSourcePath);
+            cmd.add("-n", pool._poolAuthUserName);
+            cmd.add("-s", pool._poolAuthSecret);
```
Should we not use the UUID of the host? IPs can change, UUIDs should not

wido commented on scripts/vm/hypervisor/kvm/kvmheartbeat_rbd.sh:
```diff
@@ -0,0 +1,151 @@
+create_cephConf() {
+#Creating Ceph keyring and conf for executing rados commands
+  if [ ! -f $keyringFile ]; then
+    echo -e "[client.$PoolAuthUserName]\n key=$PoolAuthSecret" > $keyringFile
```
We should never write to files in /etc/ceph from a script like this.

CloudStack hosts can work without files in /etc/ceph

wido replied:
This still stands. We should not write to /etc, never.

IF we need such a file, use /run/cloudstack/agent (for example) to make sure the files are gone after a reboot.

wido commented on scripts/vm/hypervisor/kvm/kvmheartbeat_rbd.sh:
```diff
@@ -0,0 +1,151 @@
+    for ip in $(echo $SourceHostIP | sed 's/,/ /g'); do
+        confContents+="[v2:${ip}:3300/0,v1:${ip}:6789/0], "
+    done
+    echo -e "$confContents" | sed 's/, $//' > $confFile
```
Same here. We should not write to /etc/ceph/ceph.conf in any circumstance.

Not needed as well

wido reviewed (commented):
I am not convinced that this is the best route forward.

RBD already supports watchers and exclusive locking on images by default. If a host dies you should see within 30 sec or so if it is no longer watching a RBD image and also has not refreshed it's lock anymore.

This external script we are calling also does not support IPv6 (InetAddress) and would rely on external tooling.

Imho we should have each Host maybe just 'watch' a RBD image. Other hosts can check if there still is a watcher on that image. If not we can assume the Host has died.

Overall I think we should not merge this and rely on the things Ceph already has like exclusive locking and watchers on RBD images.

sjyu1 committed:
rbd watcher 방식으로 변경

1. 각 호스트용 rbd 이미지(hb-hostip) 생성하여 watcher 감시
2. /etc/ceph 파일생성 대신 -c, -k 옵션을 사용하여 rbd 실행
3. keyring파일은 /etc/cloudstack/agent/ 경로에 생성

sjyu1 committed:
kvm ha > rbd 스크립트 poolname 추가

kvm ha > rbd 스크립트 poolname 추가

Commit:
Merge pull request #148 from sjyu1/rbd-primary-ha

kvm ha watcher방식으로 변경

dhslove commented:
@wido 

Based on your comments, I changed the heartbeat/activity check method to the rbd watcher check method.
In this commit, unnecessary code has been removed to support IPv4/6, and code changes have been applied to prevent duplicate code.
Code review please.

wido commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHABase.java:
```diff
@@ -57,6 +58,28 @@ public NfsStoragePool(String poolUUID, String poolIp, String poolSourcePath, Str
+
+        public RbdStoragePool(String poolUUID, String poolIp, String poolSourcePath, String mountDestPath, PoolType type, String poolAuthUserName, String poolAuthSecret, String poolSourceHost) {
+            _poolUUID = poolUUID;
+            _poolIp = poolIp;
```
I would not call this 'poolIP'. The correct naming is the mon_host as it can contain an IP-address of a hostname.

wido commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAChecker.java:
```diff
@@ -71,6 +73,29 @@ public Boolean checkingHeartBeat() {
+            cmd.add("-p", pool._poolMountSourcePath);
+            cmd.add("-n", pool._poolAuthUserName);
+            cmd.add("-s", pool._poolAuthSecret);
+            cmd.add("-h", hostIp);
```
I know we use this with NFS, but isn't it better to use the UUID of the host here?

Each host has a unique UUID and that seems safer to use then the Private IPv4 address of the host.

wido commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAChecker.java:
```diff
@@ -71,6 +73,29 @@ public Boolean checkingHeartBeat() {
+                    pool._poolIp));
+
+            if (result == null && parsedLine.contains("DEAD")) {
+                s_logger.warn(String.format("Checking heart beat with KVMHAChecker command [%s] returned [%s]. [%s]. It may cause a shutdown of host IP [%s].", cmd.toString(),
```
Then also use the UUID here.

wido commented on plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java:
```diff
@@ -139,7 +139,13 @@ public DataStore initialize(Map<String, Object> dsInfos) {
         URI uri = null;
+        boolean multi = false;
         try {
+            String urlType = url.substring(0, 3);
```
Should we not do this split after we have used the URI class and then just split the 'host' part?

Now you are doing manual string parsing on a string which we will then run through a library.

wido commented on plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java:
```diff
@@ -162,10 +168,18 @@ public DataStore initialize(Map<String, Object> dsInfos) {
                 if (uriPath == null) {
                     throw new InvalidParameterValueException("host or path is null, should be rbd://hostname/pool");
                 }
+                if (multi) {
```
We can probably just split uriHost here.

wido commented on scripts/vm/hypervisor/kvm/kvmheartbeat_rbd.sh:
```diff
@@ -0,0 +1,136 @@
+                    -n pool auth username
+                    -s pool auth secret
+                    -h host
+                    -i source host ip
```
Again, I suggest using the UUID of the Host

wido commented on scripts/vm/hypervisor/kvm/kvmheartbeat_rbd.sh:
```diff
@@ -0,0 +1,136 @@
+                    -s pool auth secret
+                    -h host
+                    -i source host ip
+                    -r cretae/read hb watcher
```
Typo

wido commented on scripts/vm/hypervisor/kvm/kvmheartbeat_rbd.sh:
```diff
@@ -0,0 +1,136 @@
+  exit 2
+fi
+
+keyringFile="/etc/cloudstack/agent/keyring"
```
This will break if multiple Ceph clusters are attached to the same host.

This is technically possible and also done (we do this for example).

At least you need a unique keyfile per Ceph cluster, but I would even say that we should not write this credentials to a file anyway.

If we do so at least write it to a temporary directory in /run somewhere and make sure the chmod permissions are set properly so no other user can read it.

wido commented on scripts/vm/hypervisor/kvm/kvmheartbeat_rbd.sh:
```diff
@@ -0,0 +1,136 @@
+  /usr/bin/logger -t heartbeat "kvmheartbeat_rbd.sh reboots the system because there is no heartbeat watcher."
+  sync &
+  sleep 5
+  echo b > /proc/sysrq-trigger
```
Do we really need to kill the host? Since Ceph/RBD lives in userspace we could also 'just' kill all the VMs which are using this storage pool.

No need to reboot the host completely.

With NFS you have something hanging in kernel space and causes I/O to lock up. This is not the case with RBD.

wido commented on scripts/vm/hypervisor/kvm/kvmvmactivity_rbd.sh:
```diff
@@ -0,0 +1,107 @@
+fi
+
+#Creating Ceph keyring for executing rbd commands
+keyringFile="/etc/cloudstack/agent/keyring.bin"
```
See my earlier comment about these keyring files.

wido reviewed (commented):
First: Thanks for the changes!

Overall it's an improvement, but I don't think we are there yet.

I also doubt if Bash is the proper language to do this checking. Python has great librbd bindings which even make checking the RBD watcher easier: https://docs.ceph.com/en/pacific/rbd/api/librbdpy/

And also: https://github.com/ceph/ceph/blob/master/src/pybind/rbd/rbd.pyx

<pre>watchers_list()</pre>

Also my comment in general:

- We should probably not write the RBD secrets to a file. We can just pass them as arguments
- Keep in mind multiple Ceph/RBD pools can be in use in a CloudStack setup
- Is Bash really the best language here?
- We should probably use the host UUID and not Private IPv4 address

ravening commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHABase.java:
```diff
@@ -32,9 +32,10 @@
     protected long _heartBeatUpdateTimeout = 60000;
     protected long _heartBeatUpdateFreq = 60000;
-    protected long _heartBeatUpdateMaxTries = 5;
+    protected long _heartBeatUpdateMaxTries = 10;
```
Instead of changing the existing value, is it better to use new variable for ceph?

ravening commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:
```diff
@@ -35,121 +35,189 @@
     public void addStoragePool(NfsStoragePool pool) {
-        synchronized (storagePool) {
-            storagePool.put(pool._poolUUID, pool);
+        synchronized (nfsstoragePool) {
```
Since you are already using `ConcurrentHashMap`, I dont see the need of synchronized

ravening commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:
```diff
@@ -35,121 +35,189 @@
-                        continue;
-                    }
+    public NfsStoragePool getStoragePool(String uuid) {
+        synchronized (nfsstoragePool) {
```
`synchronized` is not needed here. same case below also.

ravening commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:
```diff
@@ -35,121 +35,189 @@
             }
         }
 
+        return removedPools;
```
You are not doing anything with the returned set in line 125. So `return` is not needed

ravening commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAVMActivityChecker.java:
```diff
@@ -27,36 +28,50 @@
-        cmd.add("-p", nfsStoragePool._poolMountSourcePath);
-        cmd.add("-m", nfsStoragePool._mountDestPath);
+        String poolIp = "";
+        if (poolType == StoragePoolType.NetworkFilesystem) {
```
Rather than adding `if/else` extract them according the pool type. In future, if we support more pool type, they can goto their own functions else code will look messy

sjyu1 committed:
hb,ac 체크 스크립트를 Python librbd로 변경

sjyu1 committed:
소스복원

sjyu1 committed:
heartbeatUpdateMaxTries 수정, synchronized 삭제

sjyu1 committed:
KVMHAMonitor

sjyu1 committed:
librbd setting, mon_host to conffile

Commit:
Merge pull request #159 from sjyu1/rbd-primary-ha

hb,ac 체크 스크립트를 Python librbd로 변경

dhslove commented:
@wido @ravening 

Thank you for your comments. We took your comments and modified the code.

1. Use Python, librbd.
2. Use Ceph Watcher for hb, ac work.
3. Key file and secret are created and used temporarily, and temporary files are created in the /run/cloudstack/agent directory.

We didn't apply your comment to the code to use host uuid as the name of the hb image. In order for the HA program to use the Host uuid, the management side code needs to be changed, but we think that using the existing code is a better way.

sjyu1 committed:
ASF 라이센스 정보를 작성 및 공백제거

Commit:
Merge pull request #162 from sjyu1/rbd-primary-ha

ASF 라이센스 정보를 작성 및 공백제거

sjyu1 committed:
Changed so that ha occurs when oobm is off

sjyu1 committed:
Changed so that ha occurs when oobm is off

sjyu1 committed:
OOBM PowerState.off시 로그추가

sjyu1 committed:
주석삭제

Commit:
Merge pull request #176 from sjyu1/rbd-primary-ha

Changed so that ha occurs when oobm is off

dhslove commented:
@weizhouapache 

I checked the issue(#4211) you requested and added handling of RBD to KVMInvestigator.java.

And the fundamental problem of this issue(#4211) is that HA does not work normally when the host is in Off state. To solve this, if the host is already off in the recovering stage, it is reasonable to treat it as recover failure and then switch to the fencing stage. I solved this problem by modifying the source of KVMHAProvider.java.

sjyu1 committed:
Merge pull request #17 from ablecloud-team/rbd-primary-ha

Rbd primary ha

sjyu1 committed:
INVALIDATE CACHE Test

sjyu1 committed:
delete import

sjyu1 committed:
소스정리

sjyu1 committed:
소스원복

sjyu1 committed:
rbd image-cache invalidate

sjyu1 committed:
volume uuid -> path로 변경

Commit:
Merge pull request #182 from sjyu1/rbd-primary-ha

rbd image-cache invalidate

dhslove committed:
Revert "Merge pull request #182 from sjyu1/rbd-primary-ha"

This reverts commit 685cfbef52f566fa1a8ce0f2cdfa7b4d781f9fd8, reversing
changes made to f2c7ccb0add391837f8d4ee77351426a613016a1.

blueorangutan commented:
@acs-robot a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

blueorangutan commented:
UI build: :heavy_check_mark:
Live QA URL: http://qa.cloudstack.cloud:8080/client/pr/5862 (SL-JID-1755)

weizhouapache commented:
@dhslove good feature.

multiple ceph rbd monitors are supported in #6792 , can you update this PR to fix the conflicts ?
can you please also address Wido's outstanding comments ?

sjyu1 committed:
Fix conflicting sources

Commit:
Merge pull request #253 from sjyu1/rbd-primary-ha2

Source fixes to support multiple ceph rbd monitors

dhslove commented:
@weizhouapache Sorry for the late reply. A source conflict was fixed. Please check.

sjyu1 committed:
파이썬 버전 설정

Commit:
Merge pull request #255 from sjyu1/rbd-primary-ha2

py파일에 파이썬 버전 설정

dhslove commented on scripts/vm/hypervisor/kvm/kvmheartbeat_rbd.py:
```diff
@@ -0,0 +1,115 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
@wido I have applied your request.

dhslove commented on scripts/vm/hypervisor/kvm/kvmvmactivity_rbd.py:
```diff
@@ -0,0 +1,103 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
@wido I have applied your request.

wido reviewed (approved):
Code looks good

DaanHoogland commented:
@GutoVeronezi @ravening I resolved all outdated comments. cc @weizhouapache. Do you giys approve?

weizhouapache commented:
@blueorangutan package

blueorangutan commented:
Packaging result: :heavy_multiplication_x: el7 :heavy_multiplication_x: el8 :heavy_multiplication_x: debian :heavy_multiplication_x: suse15. SL-JID 4814

weizhouapache commented:
@dhslove 
can you solve the compilation error ?

```
[ERROR] COMPILATION ERROR : 
[ERROR] /home/travis/build/apache/cloudstack/plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java:[241,16] error: cannot find symbol
  symbol:   variable multi
  location: class CloudStackPrimaryDataStoreLifeCycleImpl
[ERROR] /home/travis/build/apache/cloudstack/plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java:[241,15] error: illegal parenthesized expression
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project cloud-plugin-storage-volume-default: Compilation failure: Compilation failure: 
[ERROR] /home/travis/build/apache/cloudstack/plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java:[241,16] error: cannot find symbol
[ERROR]   symbol:   variable multi
[ERROR]   location: class CloudStackPrimaryDataStoreLifeCycleImpl
[ERROR] /home/travis/build/apache/cloudstack/plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/CloudStackPrimaryDataStoreLifeCycleImpl.java:[241,15] error: illegal parenthesized expression
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :cloud-plugin-storage-volume-default
The command "./tools/travis/install.sh" failed and exited with 1 during .
```

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHABase.java:
```diff
@@ -57,6 +58,28 @@ public NfsStoragePool(String poolUUID, String poolIp, String poolSourcePath, Str
     }
 
+    public static class RbdStoragePool {
+        String _poolUUID;
```
IMHO, using strict camel case (non consecutive capitals/treat acronyms as normal words) would improve readability; with this approach, the variable name would be `poolUuid`. Also, this approach is already being applied in the class name: it is `RbdStoragePool` instead of  `RBDStoragePool`.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHABase.java:
```diff
@@ -57,6 +58,28 @@ public NfsStoragePool(String poolUUID, String poolIp, String poolSourcePath, Str
+        String _poolAuthSecret;
+        String _poolSourceHost;
+
+        public RbdStoragePool(String poolUUID, String monHost, String poolSourcePath, String mountDestPath, PoolType type, String poolAuthUserName, String poolAuthSecret, String poolSourceHost) {
```
The same remark about applying camel case to acronyms.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAChecker.java:
```diff
@@ -71,6 +78,46 @@ public Boolean checkingHeartBeat() {
+        for (RbdStoragePool rbdpools : rbdStoragePools) {
+            hostAndPools = String.format("host IP [%s] in pools [%s]", hostIp, rbdStoragePools.stream().map(pool -> pool._monHost).collect(Collectors.joining(", ")));
+            s_logger.debug(String.format("Checking heart beat with KVMHAChecker for %s", hostAndPools));
+
+            ProcessBuilder processBuilder = new ProcessBuilder();
+            processBuilder.command().add("python3");
+            processBuilder.command().add(s_heartBeatPathRbd);
+            processBuilder.command().add("-i");
+            processBuilder.command().add(rbdpools._poolSourceHost);
+            processBuilder.command().add("-p");
+            processBuilder.command().add(rbdpools._poolMountSourcePath);
+            processBuilder.command().add("-n");
+            processBuilder.command().add(rbdpools._poolAuthUserName);
+            processBuilder.command().add("-s");
+            processBuilder.command().add(rbdpools._poolAuthSecret);
+            processBuilder.command().add("-v");
+            processBuilder.command().add(hostIp);
+            processBuilder.command().add("-r");
+            processBuilder.command().add("r");
+            Process process = null;
+            String parsedLine = "";
+            try {
+                process = processBuilder.start();
+                BufferedReader bfr = new BufferedReader(new InputStreamReader(process.getInputStream()));
+                parsedLine = bfr.readLine();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+
+            s_logger.debug(String.format("Checking heart beat with KVMHAChecker [{command=\"%s\", log: \"%s\", pool: \"%s\"}].", processBuilder.command().toString(), parsedLine,
+            rbdpools._monHost));
+
+            if (process != null && parsedLine.contains("DEAD")) {
+                s_logger.warn(String.format("Checking heart beat with KVMHAChecker command [%s] returned. [%s]. It may cause a shutdown of host IP [%s].", processBuilder.command().toString(),
+                        parsedLine, hostIp));
+            } else {
+                validResult = true;
+            }
+        }
+
```
This block could be extracted to a method, to better organize the code.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAChecker.java:
```diff
@@ -71,6 +78,46 @@ public Boolean checkingHeartBeat() {
+            hostAndPools = String.format("host IP [%s] in pools [%s]", hostIp, rbdStoragePools.stream().map(pool -> pool._monHost).collect(Collectors.joining(", ")));
+            s_logger.debug(String.format("Checking heart beat with KVMHAChecker for %s", hostAndPools));
+
+            ProcessBuilder processBuilder = new ProcessBuilder();
+            processBuilder.command().add("python3");
+            processBuilder.command().add(s_heartBeatPathRbd);
+            processBuilder.command().add("-i");
+            processBuilder.command().add(rbdpools._poolSourceHost);
+            processBuilder.command().add("-p");
+            processBuilder.command().add(rbdpools._poolMountSourcePath);
+            processBuilder.command().add("-n");
+            processBuilder.command().add(rbdpools._poolAuthUserName);
+            processBuilder.command().add("-s");
+            processBuilder.command().add(rbdpools._poolAuthSecret);
+            processBuilder.command().add("-v");
+            processBuilder.command().add(hostIp);
+            processBuilder.command().add("-r");
+            processBuilder.command().add("r");
+            Process process = null;
+            String parsedLine = "";
+            try {
+                process = processBuilder.start();
+                BufferedReader bfr = new BufferedReader(new InputStreamReader(process.getInputStream()));
+                parsedLine = bfr.readLine();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
```
We should use the class `com.cloud.utils.script.Script` here (same as it is being doing to NFS), as it already does all the command handling.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:
```diff
@@ -35,121 +38,168 @@
         if (pool != null) {
-            storagePool.put(pool._poolUUID, pool);
+            nfsstoragePool.put(pool._poolUUID, pool);
+        }else if (rbdpool != null) {
```
```suggestion
        } else if (rbdpool != null) {
```

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:
```diff
@@ -35,121 +38,168 @@
+                result = cmd.execute();
+                s_logger.debug(String.format("The command [%s], to the pool [%s], had the result [%s].", cmd.toString(), uuid, result));
+            } else if (rbdStoragePool != null) {
+                ProcessBuilder processBuilder = createRbdHeartBeatCommand(rbdStoragePool, hostPrivateIp, true);
+                try {
+                    process = processBuilder.start();
+                    BufferedReader bfr = new BufferedReader(new InputStreamReader(process.getInputStream()));
+                    result = bfr.readLine();
+                } catch (IOException e) {
+                    e.printStackTrace();
                 }
```
Same here about using `com.cloud.utils.script.Script`.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAMonitor.java:
```diff
@@ -169,6 +219,27 @@ private Script createHeartBeatCommand(NfsStoragePool primaryStoragePool, String
     }
 
+    private ProcessBuilder createRbdHeartBeatCommand(RbdStoragePool primaryStoragePool, String hostPrivateIp, boolean hostValidation) {
+        ProcessBuilder processBuilder = new ProcessBuilder();
+        processBuilder.command().add("python3");
+        processBuilder.command().add(s_heartBeatPathRbd);
+        processBuilder.command().add("-i");
+        processBuilder.command().add(primaryStoragePool._poolSourceHost);
+        processBuilder.command().add("-p");
+        processBuilder.command().add(primaryStoragePool._poolMountSourcePath);
+        processBuilder.command().add("-n");
+        processBuilder.command().add(primaryStoragePool._poolAuthUserName);
+        processBuilder.command().add("-s");
+        processBuilder.command().add(primaryStoragePool._poolAuthSecret);
+
+        if (hostValidation) {
+            processBuilder.command().add("-v");
+            processBuilder.command().add(hostPrivateIp);
+        }
+
+        return processBuilder;
```
Same here about using `com.cloud.utils.script.Script`.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/KVMHAVMActivityChecker.java:
```diff
@@ -18,48 +18,91 @@
+            LOG.debug(String.format("Checking heart beat with KVMHAVMActivityChecker [{command=\"%s\", result: \"%s\", log: \"%s\", pool: \"%s\"}].", cmd.toString(), result, parsedLine, nfsStoragePool._poolIp));
+
+        } else if (poolType == StoragePoolType.RBD) {
+            ProcessBuilder processBuilder = new ProcessBuilder();
+            processBuilder.command().add("python3");
+            processBuilder.command().add(vmActivityCheckPath);
+            processBuilder.command().add("-i");
+            processBuilder.command().add(rbdStoragePool._poolSourceHost);
+            processBuilder.command().add("-p");
+            processBuilder.command().add(rbdStoragePool._poolMountSourcePath);
+            processBuilder.command().add("-n");
+            processBuilder.command().add(rbdStoragePool._poolAuthUserName);
+            processBuilder.command().add("-s");
+            processBuilder.command().add(rbdStoragePool._poolAuthSecret);
+            processBuilder.command().add("-v");
+            processBuilder.command().add(hostIp);
+            processBuilder.command().add("-u");
+            processBuilder.command().add(volumeUuidList);
+            command = processBuilder.command().toString();
+            Process process = null;
+
+            try {
+                process = processBuilder.start();
+                BufferedReader bfr = new BufferedReader(new InputStreamReader(process.getInputStream()));
+                parsedLine = bfr.readLine();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
```
Same here about using `com.cloud.utils.script.Script`.

GutoVeronezi commented on plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/kvm/ha/KVMHAProvider.java:
```diff
@@ -88,8 +99,13 @@ public boolean recover(Host r) throws HARecoveryException {
-                final OutOfBandManagementResponse resp = outOfBandManagementService.executePowerOperation(r, PowerOperation.OFF, null);
-                return resp.getSuccess();
+                final OutOfBandManagement oobm = outOfBandManagementDao.findByHost(r.getId());
+                if (oobm.getPowerState() == PowerState.Unknown){
```
```suggestion
                if (oobm.getPowerState() == PowerState.Unknown) {
```

GutoVeronezi reviewed (commented):
@dhslove, ACS already uses java bindings for librados in `com.ceph.rados.Rados`; could not the heartbeat be implemented through Java instead of creating Python scripts? cc: @wido

sjyu1 committed:
multi monitor 소스삭제 및 관련파일 수정

sjyu1 committed:
poolUUID를 poolUuid로 카멜케이스 적용

sjyu1 committed:
if문 space 적용

sjyu1 committed:
supportHa () 메소드 추가

Commit:
Merge pull request #272 from sjyu1/rbd-primary-ha01

rbd primary ha 추가사항 반영

shwstppr commented:
@dhslove  can you please address the merge conflicts? Should we consider this for 4.19.0.0?

weizhouapache commented:
moved to 4.21.0.0 milestone

harikrishna-patnala commented:
@dhslove can you please advice if this PR is still valid and can we target this to 4.22 ?

dhslove commented:
@harikrishna-patnala Sorry for the late reply. I think it would be better to close this PR and implement this feature in a new PR rather than maintain it. I currently have the tested code for this feature, so I think I can contribute it to the next CloudStack version. Please let me know your thoughts.

DaanHoogland commented:
> @harikrishna-patnala Sorry for the late reply. I think it would be better to close this PR and implement this feature in a new PR rather than maintain it. I currently have the tested code for this feature, so I think I can contribute it to the next CloudStack version. Please let me know your thoughts.

thanks @dhslove , that would be great, go ahead.

calvix commented:
Hello, any plans to continue working on this?

weizhouapache commented:
> Hello, any plans to continue working on this?

@calvix 
I will test it for ACS 4.24.0
